include/secp256k1_ecdsa_s2c.h

```diff
@@ -0,0 +1,180 @@
+    const unsigned char* seckey,
+    const unsigned char* s2c_data32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(4) SECP256K1_ARG_NONNULL(5) SECP256K1_ARG_NONNULL(6);
+
+/** Verify a sign-to-contract commitment.
+ *
+ *  Returns: 1: the signature contains a commitment to data32
+ *           0: incorrect opening
+ *  Args:    ctx: a secp256k1 context object, initialized for verification.
+ *  In:      sig: the signature containing the sign-to-contract commitment (cannot be NULL)
+ *        data32: the 32-byte data that was committed to (cannot be NULL)
+ *       opening: pointer to the opening created during signing (cannot be NULL)
+ */
+SECP256K1_API SECP256K1_WARN_UNUSED_RESULT int secp256k1_ecdsa_s2c_verify_commit(
+    const secp256k1_context* ctx,
+    const secp256k1_ecdsa_signature *sig,
+    const unsigned char *data32,
+    const secp256k1_ecdsa_s2c_opening *opening
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(4);
+
+
+/** ECDSA Anti-Klepto Protocol
+ *
+ *  The ecdsa_anti_klepto_* functions can be used to prevent a signing device from
+ *  exfiltrating the secret signing keys through biased signature nonces. The general
+ *  idea is that a host provides additional randomness to the signing device client
+ *  and the client commits to the randomness in the nonce using sign-to-contract.
+ *
+ *  The following scheme is described by Stepan Snirigiv here:
+ *    https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-February/017655.html
+ *  and by Pieter Wuille (as "Scheme 6") here:
+ *    https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2020-March/017667.html
+ *
+ *  In order to ensure the host cannot trick the signing device into revealing its
+ *  keys, or the signing device to bias the nonce despite the host's contributions,
+ *  the host and client must engage in a commit-reveal protocol as follows:
+ *  1. The host draws randomness `rho` and computes a sha256 commitment to it using
+ *     `secp256k1_ecdsa_anti_klepto_host_commit`. It sends this to the signing device.
+ *  2. The signing device computes a public nonce `R` using the host's commitment
+ *     as auxiliary randomness, using `secp256k1_ecdsa_anti_klepto_signer_commit`.
+ *     The signing device sends the resulting `R` to the host as a s2c_opening.
+ *  3. The host replies with `rho` generated in step 1.
+ *  4. The device signs with `secp256k1_ecdsa_s2c_sign`, using `rho` as `s2c_data`,
+ *     and sends the signature to the host. (It can drop the `s2c_opening` that this
+ *     outputs since it will match the public nonce from step 2.)
+ *  5. The host verifies that the signature's public nonce matches the opening from
+ *     step 2 and its original randomness `rho`, using `secp256k1_ecdsa_s2c_verify_commit`.
+ *     If this fails, it indicates a potentially serious problem with the hardware device,
+ *     and we strongly discourage its continued use.
```

Comment:
Perhaps `s/fails/fails or if the signing device aborts after step 3/` and `s/use./use (see Rationale below)./` With rationale

- If the hardware device is not discarded after wrongly or not at all responding after step 3, the hardware device can successfully exfiltrate the signing key _even if_ the host only lets signatures pass after verifying the commitment. This is because after step 3, the signing device knows the final nonce and can simply abort if it does not match the desired bias.

Hm, I'm really skeptical about implementers getting this right.

Reply:
Hm yeah... Idk, throwing away the device may not be a solution either if it's the only thing that stores your key.  It's just hard to give advice in that case.

We should consistently call it "signing device" or "hardware device".

Reply:
Throwing away the signer cause of one failed session might be too much, especially since there might be transient errors (e.g. a communication error after step 3). It might be good to distinguish between a stopped process after step 3, and an actually failed commitment check.

Maybe something like 'If the signing process fails repeatedly'? To produce even one signature with a desired bias, multiple attempts by the signing device might be needed, and multiple signatures with a correct bias are needed to leak a full key. 

Of course, it might fail 'repeatedly' over many days or months and for many different transactions, which is harder to recognize for a user :shrug:

Reply:
Well, failing every few weeks or months is not a big deal. In principle you can exfiltrate a nonce that way but in practice you cannot. Even if you have a whole bit of bias (and failing every time is likely to make a user throw away your device even if they don't know any crypto..) you still need hundreds of signatures with the same key to exfiltrate a nonce this way. And a single bit is on the edge of what's computationally feasible to work with.

It's probably fine to say "if this fails a noticeable number of times, over the lifetime of the device, you should throw it away".

Reply:
@jonasnick 

> If the hardware device is not discarded after wrongly or not at all responding after step 3, the hardware device can successfully exfiltrate the signing key even if the host only lets signatures pass after verifying the commitment. This is because after step 3, the signing device knows the final nonce and can simply abort if it does not match the desired bias.

It only knows the final nonce if the host nonce is the same in a restart, right? If the host picks a new random nonce contribution every time, a restart after step 2 should be safe.

So the comment in the code:

> If, at any point from this step onward, the hardware device fails, it is
> okay to restart the protocol using **exactly the same `rho`** and checking
> that the hardware device proposes **exactly the same** `R`. Otherwise, the
> hardware device may be selectively aborting and thereby biasing the set of
> nonces that are used in actual signatures.

is not complete I think. We could add that the protocol can also be restarted with a fresh `rho`. 

Am I missing something?

Reply:
Not sure I understand
1. Restart after step 2 is the same as restart after step 3, because the host must provide the same `rho` in step 3.
2. Restarting with a fresh `rho` means restarting from step 1 which allows the hardware wallet to bias the final nonce.

Reply:
> Restart after step 2 is the same as restart after step 3, because the host must provide the same rho in step 3.

How so? After step 2 (before step 3), the signer does not know the final nonce yet.

I get that after step 3, the signer knows `rho` and with that the final nonce, so it can introduce bias by aborting the process. 

But if there is a restart before that (e.g. after the signer commitment is delivered, before sending the `rho`), restarting with a fresh `rho` should be fine? 

Not that it matters much in practice, it was more for my information and to maybe move the comment in the code to step 3.

Reply:
Oh, I think I misunderstood.

> a restart after step 2 should be safe.

You mean you can restart the whole protocol if step 3 hasn't happened yet ( I thought you mean that after restart the protocol can continue after step 2). I think in that case it can also be run with the same rho.The host can just re-request step 2 if the HWW failst to provide a step 2 messge. Yeah I think it would make sense to move the "If, at any point from this step onward, the hardware device fails, it is [...]" section to step 3 but on the other hand it shouldn't make a difference in practice - at least I don't really see how the protocol can fail at step 3.

Reply:
Thanks for confirming.

> at least I don't really see how the protocol can fail at step 3.

When the host sends `rho`, the signer could just not respond, or respond with an error message, or respond with invalid data.

Reply:
> It's probably fine to say "if this fails a noticeable number of times, over the lifetime of the device, you should throw it away".

Throwing it away isn't enough.  You should act like your master secret has been compromised.